### PR TITLE
fix(components): eliminate forwardRef errors in anticipation of React v19

### PR DESCRIPTION
**Pull request by aleciavogel**
**TL;DR:** Fixes the error spam caused by `React.forwardRef` in newer versions of Next.js (v14.2 and up). See #3898 for more info.

### Problem

```
Warning: Accessing element.ref is no longer supported. ref is now a regular prop. It will be removed from the JSX Element type in a future release.
    at eval (webpack-internal:///(ssr)/./node_modules/@radix-ui/react-slot/dist/index.mjs:46:23)
    at eval (webpack-internal:///(ssr)/./node_modules/@radix-ui/react-slot/dist/index.mjs:20:23)
    at FormControl (webpack-internal:///(ssr)/./common/ui/form.tsx:99:24)
    ...
```

### Solution

[In React v19:](https://react.dev/blog/2024/04/25/react-19#ref-as-a-prop)
> New function components will no longer need `forwardRef`, and we will be publishing a [codemod](https://github.com/codemod-com/codemod/tree/main/packages/codemods/react/19/remove-forward-ref) to automatically update your components to use the new ref prop. In future versions we will deprecate and remove forwardRef.

Since there have been [issues reported with the codemod](https://www.reddit.com/r/nextjs/comments/1dtukjs/migrate_existing_edited_shadcn_components_to/), I've manually made all the necessary changes. Now, the `ref` is automatically typed as `React.LegacyRef` through my use of `React.ComponentProps`.

#### Key Changes

- **Eliminated Redundancy**: Removed the use of `forwardRef` and used `React.ComponentProps` for automatic typing of `ref`. This simplifies component definitions and will hopefully make this codebase more approachable for TypeScript newbies.
- **Consistency and Flexibility**: Refactored components to use `React.FC` and `React.ComponentProps` for consistency and to facilitate easier use of refs for components that previously weren't using `forwardRef`.
- **Removed Redundant `ref` Prop**: Since `ref` is now included as a prop and `{...props}` spreads it, I removed the explicit `ref={ref}` line from most components. For example:
  
  **Before:**
  ```typescript
  const CardHeader = React.forwardRef<
    HTMLDivElement,
    React.HTMLAttributes<HTMLDivElement>
  >(({ className, ...props }, ref) => (
    <div
      ref={ref}
      className={cn("flex flex-col space-y-1.5 p-6", className)}
      {...props}
    />
  ));
  ```

  **After:**
  ```typescript
  const CardHeader: React.FC<React.ComponentProps<"div">> = ({
    className,
    ...props
  }) => (
    <div
      className={cn("flex flex-col space-y-1.5 p-6", className)}
      {...props}
    />
  );
  ```

- **TailwindCSS Classname Order**: Fixed the order of TailwindCSS classnames per the eslint config.
- **Spelling and Type Corrections**: Corrected spelling errors and incorrect types (e.g., using `HTMLParagraphElement` for `<h3 />` elements).

### Additional Notes

- You will not need to change how the components are being used. They should all still work as before. All tests were passing before I made this PR.
- I cannot confirm if all warnings are eliminated since `radix` packages might still need updates. However, in my personal project using `Button`, `Menubar` and `Input`, these changes removed the warnings.
- I ran `build:registry` but noticed the `__registry__` folder wasn't updated. Please let me know if I need to manually update this as well.

### Request for Feedback

This is my first contribution to this repo, so please don't hesitate to let me know if I've missed anything or if any changes need further granularity. I'm happy to squash commits if needed.

### CODE REVIEW

File: apps/www/registry/default/ui/breadcrumb.tsx [hunk 1]
```diff
@@ -102,7 +96,7 @@ const BreadcrumbEllipsis = ({
     <span className="sr-only">More</span>
   </span>
 )
-BreadcrumbEllipsis.displayName = "BreadcrumbElipssis"
+BreadcrumbEllipsis.displayName = "BreadcrumbEllipsis"
```
Comment (aleciavogel) [inline apps/www/registry/default/ui/breadcrumb.tsx:99]: This spelling mistake was present in both styles. Not sure if it was intentional, but I went ahead and corrected it on the off-chance it wasn't.

File: apps/www/registry/default/ui/breadcrumb.tsx [hunk 2]
```diff
@@ -88,10 +82,10 @@ const BreadcrumbSeparator = ({
 )
 BreadcrumbSeparator.displayName = "BreadcrumbSeparator"
 
-const BreadcrumbEllipsis = ({
+const BreadcrumbEllipsis: React.FC<React.ComponentProps<"span">> = ({
   className,
   ...props
-}: React.ComponentProps<"span">) => (
+}) => (
```
Comment (aleciavogel) [inline apps/www/registry/default/ui/breadcrumb.tsx:88]: An example of a component that wasn't previously using `forwardRef` that I've updated to use `React.FC` for the sake of consistency.

File: apps/www/registry/default/ui/breadcrumb.tsx [hunk 3]
```diff
@@ -4,79 +4,73 @@ import { ChevronRight, MoreHorizontal } from "lucide-react"
...
   }
->(({ asChild, className, ...props }, ref) => {
+> = ({ asChild, className, ...props }) => {
   const Comp = asChild ? Slot : "a"
 
   return (
     <Comp
-      ref={ref}
-      className={cn("transition-colors hover:text-foreground", className)}
+      className={cn("hover:text-foreground transition-colors", className)}
```
Comment (aleciavogel) [inline apps/www/registry/default/ui/breadcrumb.tsx:48]: Here is an example of a place where I removed the `ref={ref}` line, as the `ref` is now included within `{...props}` as a `React.LegacyRef` type.

File: apps/www/registry/default/ui/skeleton.tsx [hunk 4]
```diff
@@ -1,15 +1,15 @@
+import * as React from "react"
```
Comment (aleciavogel) [inline apps/www/registry/default/ui/skeleton.tsx:1]: A couple of components were missing this import (I can't recall offhand what the other one was). I've added the import for the sake of consistency (and also to eliminate the squiggly lines I get in WebStorm for using a UMD global variable 🤪)

File: apps/www/registry/default/ui/card.tsx [hunk 5]
```diff
@@ -2,78 +2,60 @@ import * as React from "react"
...
 
-const CardTitle = React.forwardRef<
-  HTMLParagraphElement,
-  React.HTMLAttributes<HTMLHeadingElement>
->(({ className, ...props }, ref) => (
+const CardTitle: React.FC<React.ComponentProps<"h3">> = ({
+  className,
+  ...props
+}) => (
   <h3
```
Comment (aleciavogel) [inline apps/www/registry/default/ui/card.tsx:31]: This is not the only component that was originally using a paragraph element for headings. To be fair, it doesn't make THAT huge of a difference, but it was gonna bother me if I didn't fix it 😅

**Comment by shadcn** [comment]
@aleciavogel I'm reviewing. I've been working on this as well. If it's okay, can I push some of my updates to your PR? Thanks for your work on this. Appreciate it.

**Comment by aleciavogel** [comment]
@shadcn of course!

**Comment by itsMapleLeaf** [comment]
I was just thinking about this, glad to see this in a PR!

I wanna mention that `React.FC` is not needed, and I'd recommend against using in this PR. See here for reasoning:

- https://github.com/facebook/create-react-app/pull/8177
- https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components/ (under "Why is `React.FC` not needed?" also with a link to a codemod to remove it)

**Comment by hsavit1** [comment]
> ## Merge state
> [ Add your review ](/shadcn-ui/ui/pull/4356/files#submit-review)

any updates on this?

**Comment by hsavit1** [comment]
> @aleciavogel I'm reviewing. I've been working on this as well. If it's okay, can I push some of my updates to your PR? Thanks for your work on this. Appreciate it.

any updates on this?

**Comment by aleciavogel** [comment]
Saw the update that shadcn is now compatible with Next 15 and therefore React v19rc so I figure my PR is no longer needed

**Comment by evheniydan** [comment]
> Saw the update that shadcn is now compatible with Next 15 and therefore React v19rc so I figure my PR is no longer needed

But shadcn components still have forwardRef. So why this PR is no longer needed? Not sure if I'm missing something here.

**Comment by aleciavogel** [comment]
> But shadcn components still have forwardRef. So why this PR is no longer needed? Not sure if I'm missing something here.

Ah, I see. I'll reopen, but my branch is likely stale so I've got some work to do on this (and someone mentioned removing `FC` so that needs to be done as well).

**Comment by waynebrantley** [comment]
@shadcn Wonder if you could provide guidance about the plan for this pr.  I would imagine there is potentially the need to support react 19 and prior to react 19.  The situation appears to be shadcn puts in react <19 compatible components and developers hand update them.

Wonder if a 'codemod' pipeline could be plugged into the cli - so when you add it runs what is added through the codemod.  This potentially could let something like this happen (as well as other items going forward)?   

Just looking for guidance

**Comment by Jacksonmills** [comment]
> I was just thinking about this, glad to see this in a PR!
> 
> I wanna mention that `React.FC` is not needed, and I'd recommend against using in this PR. See here for reasoning:
> 
>     * [Remove React.FC from Typescript template facebook/create-react-app#8177](https://github.com/facebook/create-react-app/pull/8177)
> 
>     * https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components/ (under "Why is `React.FC` not needed?" also with a link to a codemod to remove it)

Was going to mention the same thing 👌

I recommend something like this:
```
const CardHeader = ({
  className,
  ...props
}: React.ComponentProps<"div">) => (
  <div
    className={cn("flex flex-col space-y-1.5 p-6", className)}
    {...props}
  />
);
```

You can always make a `CardHeaderProps` type if needed of course

**Comment by shadcn** [comment]
@aleciavogel This is now in latest. Thanks for this PR. Appreciate your work here.
